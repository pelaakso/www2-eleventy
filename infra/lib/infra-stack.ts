import * as cdk from 'aws-cdk-lib';
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as apigw from '@aws-cdk/aws-apigatewayv2-alpha';
import * as apigwint from '@aws-cdk/aws-apigatewayv2-integrations-alpha';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as lambdaNode from 'aws-cdk-lib/aws-lambda-nodejs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as route53targets from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

export class InfraStack extends cdk.Stack {
  public readonly webSiteBucket: s3.IBucket;
  public readonly webSiteARecord: route53.ARecord;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const recordName = 'www2-eleventy';
    const domainName = 'petey952.be';
    const apiCertificateArn = 'arn:aws:acm:eu-central-1:140966923789:certificate/c96ba383-ae37-49e3-a8ea-a58eb96da369';
    const bucketName = `${recordName}.${domainName}`;
    const apiRecordName = `${recordName}-api`;
    const apiGwCustomDomainNameRecord = `${apiRecordName}.${domainName}`;
    const corsAllowedOrigins = [
      'http://localhost:3000', // for prettier: thanks for nothing
      'http://localhost:8080',
      `http://${recordName}.${domainName}`,
    ];
    const corsAllowedHeaders = [
      'Authorization', //
      'Content-Type',
    ];

    this.webSiteBucket = new s3.Bucket(this, 'Www2EleventyWebBucket', {
      autoDeleteObjects: false,
      bucketName,
      publicReadAccess: true,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
      websiteErrorDocument: '404.html',
      websiteIndexDocument: 'index.html',
    });

    const hostedZone = route53.HostedZone.fromLookup(this, 'Zone', { domainName });

    this.webSiteARecord = new route53.ARecord(this, 'Www2EleventyAliasRecord', {
      zone: hostedZone,
      recordName,
      target: route53.RecordTarget.fromAlias(new route53targets.BucketWebsiteTarget(this.webSiteBucket)),
    });

    const subscribeNewsletterFn = new lambdaNode.NodejsFunction(this, 'Www2EleventySubscribeNewsletter', {
      architecture: lambda.Architecture.ARM_64,
      entry: 'src/functions/subscribeNewsletter.ts',
      handler: 'lambdaHandler',
      runtime: lambda.Runtime.NODEJS_16_X,
      description: 'Www2Eleventy: handler for subscribe to newsletter',
      functionName: 'Www2EleventySubscribeNewsletter',
      reservedConcurrentExecutions: 10,
      timeout: cdk.Duration.seconds(15),
    });

    const subscribeNewsletterIntegration = new apigwint.HttpLambdaIntegration(
      'subscribeNewsletterIntegration',
      subscribeNewsletterFn,
      {
        payloadFormatVersion: apigw.PayloadFormatVersion.VERSION_2_0,
      }
    );

    const apiGwCustomDomainName = new apigw.DomainName(this, 'Www2EleventyApiGwCustomDomainName', {
      domainName: apiGwCustomDomainNameRecord,
      certificate: acm.Certificate.fromCertificateArn(this, 'Www2EleventyImportedCertificate', apiCertificateArn),
    });

    const httpApi = new apigw.HttpApi(this, 'Www2EleventyHttpApi', {
      apiName: `Www2EleventyBackend`,
      corsPreflight: {
        allowCredentials: true,
        allowHeaders: corsAllowedHeaders,
        allowMethods: [
          apigw.CorsHttpMethod.GET, //
          apigw.CorsHttpMethod.POST,
          apigw.CorsHttpMethod.OPTIONS,
        ],
        allowOrigins: corsAllowedOrigins,
        maxAge: cdk.Duration.minutes(5),
      },
      defaultDomainMapping: {
        domainName: apiGwCustomDomainName,
      },
      description: 'Api Gateway for Www2Eleventy backend',
    });

    httpApi.addRoutes({
      path: '/v1/subscribe-newsletter',
      integration: subscribeNewsletterIntegration,
      methods: [apigw.HttpMethod.POST],
    });

    new route53.ARecord(this, 'Www2EleventyApiAliasRecord', {
      zone: hostedZone,
      recordName: apiRecordName,
      target: route53.RecordTarget.fromAlias(
        new route53targets.ApiGatewayv2DomainProperties(
          apiGwCustomDomainName.regionalDomainName,
          apiGwCustomDomainName.regionalHostedZoneId
        )
      ),
    });

    /* To be done later.
    const role = new iam.Role(this, 'Rooli', {
      assumedBy: new iam.OpenIdConnectPrincipal(provider),
    });
    */
  }
}
