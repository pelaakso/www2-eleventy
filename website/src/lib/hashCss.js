const glob = require('fast-glob');
const hashing = require('./hashing');

function hashCssFilesSync(dir) {
    const cssFiles = glob.sync([`${dir}/**/*.css`]);
    return hashing.generateHashOfFilesSync(cssFiles);
};

exports.hashCssFilesSync = hashCssFilesSync;
