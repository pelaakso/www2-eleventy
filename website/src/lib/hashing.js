const fs = require('fs');
const crypto = require('crypto');

async function generateHashOfFiles(files) {
    const hash = crypto.createHash('md5');

    for (const file of files) {
        hash.update(await fs.promises.readFile(file));
    }

    return hash.digest('hex');
};

function generateHashOfFilesSync(files) {
    const hash = crypto.createHash('md5');

    for (const file of files) {
        hash.update(fs.readFileSync(file));
    }

    return hash.digest('hex');
};

function generateHashOfInput(str) {
    return crypto.createHash('md5')
            .update(str)
            .digest('hex');
};

exports.generateHashOfFiles = generateHashOfFiles;
exports.generateHashOfFilesSync = generateHashOfFilesSync;
exports.generateHashOfInput = generateHashOfInput;
