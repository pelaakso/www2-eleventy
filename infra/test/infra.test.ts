import { Template } from 'aws-cdk-lib/assertions';
import * as cdk from 'aws-cdk-lib';
import * as Infra from '../lib/infra-stack';

test('Whole stack snapshot', () => {
  const stack = new cdk.Stack();

  const myStack = new Infra.InfraStack(stack, 'MyTestStack', {
    env: { account: '140966923789', region: 'eu-central-1' },
  });

  const template = Template.fromStack(myStack);
  expect(template).toMatchSnapshot();
});
