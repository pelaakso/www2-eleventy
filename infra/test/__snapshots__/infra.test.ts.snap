// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Whole stack snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Www2EleventyAliasRecord96528953": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": "s3-website.eu-central-1.amazonaws.com",
          "HostedZoneId": "Z21DNDUVLTQW6Q",
        },
        "HostedZoneId": "DUMMY",
        "Name": "www2-eleventy.petey952.be.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "Www2EleventyApiAliasRecordB1B79CC2": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "Www2EleventyApiGwCustomDomainName9E5BC894",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::GetAtt": Array [
              "Www2EleventyApiGwCustomDomainName9E5BC894",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "www2-eleventy-api.petey952.be.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "Www2EleventyApiGwCustomDomainName9E5BC894": Object {
      "Properties": Object {
        "DomainName": "www2-eleventy-api.petey952.be",
        "DomainNameConfigurations": Array [
          Object {
            "CertificateArn": "arn:aws:acm:eu-central-1:140966923789:certificate/c96ba383-ae37-49e3-a8ea-a58eb96da369",
            "EndpointType": "REGIONAL",
          },
        ],
      },
      "Type": "AWS::ApiGatewayV2::DomainName",
    },
    "Www2EleventyHttpApi7C0C73DA": Object {
      "Properties": Object {
        "CorsConfiguration": Object {
          "AllowCredentials": true,
          "AllowHeaders": Array [
            "Authorization",
            "Content-Type",
          ],
          "AllowMethods": Array [
            "GET",
            "POST",
            "OPTIONS",
          ],
          "AllowOrigins": Array [
            "http://localhost:3000",
            "http://localhost:8080",
            "http://www2-eleventy.petey952.be",
          ],
          "MaxAge": 300,
        },
        "Description": "Api Gateway for Www2Eleventy backend",
        "Name": "Www2EleventyBackend",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "Www2EleventyHttpApiDefaultStage685CF17A": Object {
      "DependsOn": Array [
        "Www2EleventyApiGwCustomDomainName9E5BC894",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "Www2EleventyHttpApi7C0C73DA",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "Www2EleventyHttpApiDefaultStageDefaultMyTestStackWww2EleventyApiGwCustomDomainNameundefinedEC10E157": Object {
      "DependsOn": Array [
        "Www2EleventyApiGwCustomDomainName9E5BC894",
        "Www2EleventyHttpApiDefaultStage685CF17A",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "Www2EleventyHttpApi7C0C73DA",
        },
        "DomainName": Object {
          "Ref": "Www2EleventyApiGwCustomDomainName9E5BC894",
        },
        "Stage": "$default",
      },
      "Type": "AWS::ApiGatewayV2::ApiMapping",
    },
    "Www2EleventyHttpApiPOSTv1subscribenewsletter8A9A85F4": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "Www2EleventyHttpApi7C0C73DA",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "POST /v1/subscribe-newsletter",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "Www2EleventyHttpApiPOSTv1subscribenewsletterHttpIntegrationf43c2ff475f8ba73dc8ed1de4b5e7313FEE92118",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "Www2EleventyHttpApiPOSTv1subscribenewsletterHttpIntegrationf43c2ff475f8ba73dc8ed1de4b5e7313FEE92118": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "Www2EleventyHttpApi7C0C73DA",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "Www2EleventySubscribeNewsletterBA14B32E",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "Www2EleventyHttpApiPOSTv1subscribenewsletterMyTestStackWww2EleventyHttpApiPOSTv1subscribenewsletter884EB522PermissionDC9C421A": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Www2EleventySubscribeNewsletterBA14B32E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:eu-central-1:140966923789:",
              Object {
                "Ref": "Www2EleventyHttpApi7C0C73DA",
              },
              "/*/*/v1/subscribe-newsletter",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "Www2EleventySubscribeNewsletterBA14B32E": Object {
      "DependsOn": Array [
        "Www2EleventySubscribeNewsletterServiceRole7A916BAB",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-140966923789-eu-central-1",
          "S3Key": "069b252cbc16fbc274954a16917b809950e34ee443474d3e3dd6579e3a1f3f10.zip",
        },
        "Description": "Www2Eleventy: handler for subscribe to newsletter",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "Www2EleventySubscribeNewsletter",
        "Handler": "index.lambdaHandler",
        "ReservedConcurrentExecutions": 10,
        "Role": Object {
          "Fn::GetAtt": Array [
            "Www2EleventySubscribeNewsletterServiceRole7A916BAB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "Www2EleventySubscribeNewsletterServiceRole7A916BAB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Www2EleventyWebBucket6F46ADF8": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "www2-eleventy.petey952.be",
        "WebsiteConfiguration": Object {
          "ErrorDocument": "404.html",
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "Www2EleventyWebBucketPolicyEBD8B99C": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "Www2EleventyWebBucket6F46ADF8",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "Www2EleventyWebBucket6F46ADF8",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
